lib/rom/repository.rb

```diff
@@ -19,67 +19,35 @@ class Repository
     # @api public
     attr_reader :connection
 
-    # Additional options hash
-    #
-    # @return [Hash]
-    #
     # @api public
-    attr_reader :options
-
-    # @api private
-    def self.inherited(klass)
-      Repository.registered.unshift(klass)
+    def self.setup(repository_or_scheme, *args)
+      case repository_or_scheme
+      when String
+        fail ArgumentError, <<-STRING.gsub(/^ {10}/, '')
+          URIs without an explicit scheme are not supported anymore.
+          See https://github.com/rom-rb/rom/blob/master/CHANGELOG.md
+        STRING
+      when Symbol
+        class_from_symbol(repository_or_scheme).new(*args)
+      else
+        if args.empty?
+          repository_or_scheme
+        else
+          fail ArgumentError, "Can't accept arguments when passing an instance"
+        end
+      end
     end
 
     # @api private
-    def self.registered
-      @__registered__ ||= []
-    end
-
-    # Setup a repository instance with the given connection URI
-    #
-    # @example
-    #
-    #   Repository = Class.new(ROM::Repository)
-    #
-    #   repository = Repository.new('mysql://localhost/test')
-    #
-    #   repository.uri.scheme # => 'mysql'
-    #   repository.uri.host # => 'localhost'
-    #   repository.uri.path # => '/test'
-    #
-    # @param [String] uri_string
-    #
-    # @return [Repository]
-    #
-    # @api public
-    def self.setup(uri_string, options = {})
-      uri = Addressable::URI.parse(uri_string)
-      klass = self[uri.scheme]
-
-      unless klass
-        raise ArgumentError, "#{uri_string.inspect} uri is not supported"
+    def self.class_from_symbol(type)
+      begin
+        require "rom/#{type}"
+      rescue LoadError
+        fail ArgumentError, "#{type.inspect} is not supported"
```

Comment:
I guess it'd be nicer to have an explanation like `Failed to load rom-foo adapter`. This can happen when you simply forgot to add rom-foo to Gemfile etc.